### PR TITLE
RichText: try alternative list shortcuts (to tab)

### LINKED ISSUE #9611: Keyboard shortcuts for list items are missing
Similar issues filed: #4181, #5087, #7051, #12142.

While editing a post in Gutenberg, I couldn't figure out how to indent list items by the keyboard which is a habit of mine (and I'd argue many others editors as well). 

**Describe the bug**
No keyboard shortcuts display for the list block. 

Mentioned in https://github.com/WordPress/gutenberg/issues/7051 but that issue is not dedicated to this, I decided to make a new issue for this; close if this is too explicit. 

**To Reproduce**
without looking on a separate page of documentation, right now: 

1. add a new page or edit a page in Gutenberg
2. add a list block, 
3. try to indent a list item using only the keyboard; did ya have to look at the docs :D. 

**Expected behavior**
Personally, the expected action to indent an item in a list block would be done by issuing 'TAB' but that has been [decided against for accessibility reasons](https://github.com/WordPress/gutenberg/issues/4181#issuecomment-354091113)
That the default behavior isn't used stresses the utility of having the keyboard shortcuts for list actions (outdenting, indenting) available. 

**Screenshots**
see https://user-images.githubusercontent.com/1682452/40778281-f1054ca2-64d0-11e8-953f-704ef1e5ba1d.png

**Additional context**
- (Update: still present in WordPress 5.0.2) 
(Issue made on CPL time)

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

Fixes #9611. ~~Still needs tests.~~

I'm following up on the idea that @mcsf had regarding list indentation shortcuts in https://github.com/WordPress/gutenberg/issues/9611#issuecomment-454048403.

* `BACKSPACE` to remove indentation. This makes a TON of sense to me. It feels very intuitive.
* `SPACE` at the start to indent a list item. I like this!
  * Visually you want to *add* indentation (space), so it makes sense to use the space key for it.
  * It's a bit similar to Markdown, but then only one space.
  * I like how `SPACE` is the opposite of `BACKSPACE`.
  * Downside: it's hard to rewire the `TAB` habit, but I feel already used to it. On the other hand, it's much easier to remember than the complicated modifier shortcuts.
  * Downside: you cannot add spaces to the start of a line, but why would you? :)

Question: should we add any tooltips? If so, what symbols or words should we use?

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

**Comment by kjellr** [comment]
I gave this a try, and once you get the hang of it, it's fairly intuitive. 

- `Backspace` to outdent is super simple and makes sense — I think we should include this either way. 
- After reading this thread, `Space` to indent was really simple. I can't think of a great use case for starting a list item with a space, so it seems fine to reserve it for this. That said, I ran this by my wife with no background, and she was unable to figure out how to indent at all. I don't think many users will stumble upon this behavior, but once they do figure it out, it seems fairly easy to get used to. 

One other thing to note: On a mac, `control-option-tab` is an OS-standard shortcut for inserting a tab into a text field without tabbing out of it. I was sort of surprised to see that doesn't actually work to indent here. Might be worth adding that too?

**Comment by ellatrix** [comment]
> One other thing to note: On a mac, control-option-tab is an OS-standard shortcut for inserting a tab into a text field without tabbing out of it. I was sort of surprised to see that doesn't actually work to indent here. Might be worth adding that too?

Sure, we could include this! I wonder if there's anything like this on Windows/Linux.

**Comment by chrisvanpatten** [comment]
I love the idea! I do wonder what would happen if we did this but with 4 spaces, instead of just 1? That mirrors markdown a little more closely. Obviously means slightly extra effort on the part of the typer, but maybe a bit more intuitive for anyone coming from markdown.

Just a thought — shouldn't unblock. Maybe just worth a test for the ergonomics :)

**Comment by ellatrix** [comment]
@chrisvanpatten I had a similar idea at first, but it probably a bit harder to discover and explain. Also a bit harder to make it work. :)

**Comment by afercia** [comment]
> On a mac, control-option-tab is an OS-standard shortcut for inserting a tab into a text field 

To my knowledge, works only with Safari. Nothing happens with other browsers. Not arguing, but I wouldn't call it OS-standard 😉

**Review by afercia — APPROVED** [review]
Quickly tested: no objections from an accessibility perspective! Tab native behavior is preserved, the alternate shortcuts still work.

Noticed some weirdness with Undo/Redo but I guess they're pre-existing issues of the list block?

P.S.
Should be tested on Windows, especially with IE11.

**Comment by mcsf** [comment]
@ellatrix, I really like this, thanks for exploring!

- As you point out, <kbd>Backspace</kbd> feels exceptionally natural.
- <kbd>Space</kbd> is less obvious, but also one of our only choices. Sure, I think we can extend it to other tab-like signals (per the observation about <kbd>Ctrl+Opt+Tab</kbd>), but keep <kbd>Space</kbd> as the canonical one.
- Once the shortcut is discovered, it's quite easy to get used to it, I would say. Thus, we can get this in and then explore how to improve discoverability.
- I do notice that the shortcut tooltip is very awkward when the shortcut is <kbd>Space</kbd>:

Outdent | Indent
-|-
screenshot 2019-03-08 at 18 51 46 | screenshot 2019-03-08 at 18 51 53

**Comment by mcsf** [comment]
> * I do notice that the shortcut tooltip is very awkward when the shortcut is Space:

I realise from the diff that this is intentional. And I guess it's a short and visual way to describe the shortcut? I have no idea how we'd improve on it :), except for leaving it in plain language. Anyway, that shouldn't block this PR!

**Comment by kjellr** [comment]
> To my knowledge, works only with Safari. Nothing happens with other browsers. Not arguing, but I wouldn't call it OS-standard 😉

Oh that's odd — my mistake. You're right, it doesn't seem to work in Firefox or Chrome. Probably doesn't make sense to include in that case.

**Review by mcsf — CHANGES_REQUESTED** [review]
I was trying to understand the logic around `if ( isCollapsed( value ) && formats[ index ] && formats[ index ].length )`, which seemed odd to me, and I ran into this:

backspace-outdent-bug

^ In case it's not clear: after applying bold across those two lines, move the caret to the start of the second character of the second line, and press <kbd>Backspace</kbd>.

I think a couple of judicious comments (and/or some reduction of `else` clauses or `if/else` nesting) could help understand the flow of this.

**Comment by ellatrix** [comment]
@mcsf That problem should go away after we merge #13948, which I'd appreciate a review on. :)

**Review by mcsf — APPROVED** [review]

**Comment by ellatrix** [comment]
> I think a couple of judicious comments (and/or some reduction of else clauses or if/else nesting) could help understand the flow of this.

I added some comments in 2888652.

> ^ In case it's not clear: after applying bold across those two lines, move the caret to the start of the second character of the second line, and press Backspace.

This should be gone after the merge of #13948. I rebased this PR.

> I realise from the diff that this is intentional. And I guess it's a short and visual way to describe the shortcut? I have no idea how we'd improve on it :), except for leaving it in plain language. Anyway, that shouldn't block this PR!

I changed it to use the words "Space" and "Backspace".

Finally, an e2e has also been added.

**Comment by ellatrix** [comment]
> Noticed some weirdness with Undo/Redo but I guess they're pre-existing issues of the list block?

@afercia What issues are you experiencing? Could you create a separate issue?

**Comment by afercia** [comment]
@ellatrix sure will do: it's when nesting by multiple levels multiple list items, then undoing and redoing doesn't always restore the list to the previous correct state.

Important note: when changing keyboard shortcuts, would you please make a list of the changes and communicate them to @abrightclearweb please? She's maintaining a list that the accessibility team will publish in the accessibility docs. Thanks!

**Comment by ellatrix** [comment]
@afercia @abrightclearweb There's none removed here, we only added space to indent and backspace to outdent. For these, the caret must be at the start of a list item.

**Comment by danielbachhuber** [comment]
>     * That said, I ran this by my wife without providing any hints, and she was unable to figure out how to indent at all. I don't think many users will stumble upon this behavior, but once they do figure it out, it seems fairly easy to get used to.

I had to find this pull request in order to figure out how to indent a list item by keyboard 😆

**Pull request merged**
